src/semidiscretization.jl

```diff
@@ -0,0 +1,332 @@
+
+"""
+    ndofs(semi::AbstractSemidiscretization)
+
+Return the number of degrees of freedom associated with each scalar variable.
+"""
+@inline function ndofs(semi::AbstractSemidiscretization)
+  mesh, _, solver, cache = mesh_equations_solver_cache(semi)
+  ndofs(mesh, solver, cache)
+end
+
+
+"""
+    integrate(func, semi::AbstractSemidiscretization, u_ode::AbstractVector, args...; normalize=true)
+
+Call `func(u, i..., element, equations, solver, args...)` for all nodal indices `i..., element`
+and integrate the result using a quadrature associated with the semidiscretization `semi`.
+
+If `normalize` is true, the result is divided by the total volume of the computational domain.
+"""
+function integrate(func, semi::AbstractSemidiscretization, u_ode::AbstractVector, args...; normalize=true)
+  mesh, equations, solver, cache = mesh_equations_solver_cache(semi)
+
+  u = wrap_array(u_ode, mesh, equations, solver, cache)
+  integrate(func, mesh, equations, solver, cache, u, args..., normalize=normalize)
+end
+
+"""
+    integrate([func=(u_node,equations)->u_node,] u_ode::AbstractVector, semi::AbstractSemidiscretization; normalize=true)
+
+Call `func(u_node, equations)` for each vector of nodal variables `u_node` in `u_ode`
+and integrate the result using a quadrature associated with the semidiscretization `semi`.
+
+If `normalize` is true, the result is divided by the total volume of the computational domain.
+"""
+function integrate(func, u_ode::AbstractVector, semi::AbstractSemidiscretization; normalize=true)
+  mesh, equations, solver, cache = mesh_equations_solver_cache(semi)
+
+  u = wrap_array(u_ode, mesh, equations, solver, cache)
+  integrate(func, u, mesh, equations, solver, cache, normalize=normalize)
+end
```

Comment:
I basically copied the signatures from the old style. The second method is definitely fine for me, since it follows the same structure of `sum(func, u, context)`. We might ername the first kind of method, though - maybe `integrate_via_indices` or something similar? If we rename it, we should also adapt the order of basic arguments to match `integrate`.

Reply:
I am not sure if `integrate_via_indices` is the best name, but after thinking about it a few days have not come up with a better one. So unless the rest of @trixi-framework/principal-developers has an idea, I propose we go with it. And yes, in that case we can also adapt the order of arguments to match `integrate` to reduce confusion.

Reply:
Done in 23cb205.